File: LeetCode/Hard/0042-trapping-rain-water/0042-trapping-rain-water.go
func trap(height []int) int {
    //mlsf = max left so far and mrsf = max right so far
	water, left, right := 0, 0, len(height)-1
    mlsf, mrsf := height[left], height[right]
	for left < right {
		if height[left] < height[right] {
            if height[left]>mlsf{
                mlsf = height[left]
            }else{
                water += mlsf - height[left]
            }
            left++
		}else{
            if height[right]>mrsf{
                mrsf = height[right]
            }else{
                water += mrsf - height[right]
            }
            right-- 
        }
	}
    return water
}

File: LeetCode/Hard/0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int l = 0, r = height.size() - 1;
        int waterStored = 0;
        int mlsf = height[l];
        int mrsf = height[r];
            while (l < r) {
            if (mlsf< mrsf) {
                    l++;
                    mlsf = max(mlsf, height[l]);
                    waterStored += mlsf - height[l];
                }
            else {
                r--;
                mrsf = max(mrsf, height[r]);
                waterStored += mrsf - height[r];
            }
        }
        return waterStored;
    }
};